### DISCUSSION TITLE
How do I make `bat` use `LF` and `^D` on Windows, similar to `cat` on Linux instead of `CR LF` and `^Z` ?

### DISCUSSION THREAD

**Discussion by 500-internal-server-error** [sharkdp/bat#2884] [Q&A]
The app description says "A cat(1) clone", so I expected it to behave like `cat` more :)

GNU coreutils `cat`, whether natively on Linux or through emulation such as Cygwin/MSYS2 on Windows, _always_ uses `LF` characters, and when doing `cat | hexdump -Cv`, it waits for `^D` before terminating.

On Linux, `bat` behaves as I expect, using `LF` and waiting for `^D` when `cat`ing to `hexdump`. On Windows, however, it uses `CRLF` and waits for `^Z` instead.

I doubt there is a setting somewhere to change this, but it would be great if I could be pointed to where to patch my local install of `bat` to behave more like GNU coreutils `cat` on Windows.

Thanks in advance.

**Comment by wellkilo** [comment]
These are Windows console-runtime conventions, not a bat formatting option. In a native Windows console, Ctrl-Z is the traditional EOF input character, while Ctrl-D is handled as EOF by Unix-like terminal layers (MSYS2/Cygwin/WSL). Text-mode output may also translate LF to CRLF.

For Unix cat semantics on Windows, run bat in a Unix-like environment (WSL/MSYS2/Cygwin) and pipe within that environment. Native cmd.exe/PowerShell cannot be made to treat Ctrl-D exactly like a Unix terminal merely through a bat flag.

For file input, note that bat preserves the line structure it reads and its normal output path uses Rust stdout; there is no current --line-ending or --eof-key setting. If you need byte-for-byte concatenation, bat is also the wrong layer because syntax highlighting/decorations are text processing—use a binary-mode copy/cat implementation.

If patching a native Windows build, the relevant change belongs at process startup: call the MSVCRT _setmode(_fileno(stdin), _O_BINARY) and the same for stdout before Rust I/O wraps the handles. That disables CRLF/Ctrl-Z text translation, but Ctrl-D still depends on the terminal/shell generating EOF rather than a character; Windows console input semantics may still require Ctrl-Z or closing the pipe.

Current bat has no _setmode/O_BINARY path in its source and simply wraps io::stdin()/io::stdout() in its controller/output modules: [output.rs](https://github.com/sharkdp/bat/blob/7323a7514f7601737640e7172be115127d6db08c/src/output.rs#L202-L228), [controller.rs](https://github.com/sharkdp/bat/blob/7323a7514f7601737640e7172be115127d6db08c/src/controller.rs#L103-L119).
